Describe recursive hierarchies (#1770)
cap-js/docs: advanced/odata.md

   /ancestors(..., ID, filter(contains(Name, 'New York')), keep start)
```

#### Modeling Recursive Hierarchies

> Recursive are parent-child hierarchies, where each entity references its' parent, defining the hierarchical structure. The common example is company organization structure or HR reporting, where each employee entity references another person a as direct report or manager.

Reviewer: ```suggestion
Recursive are parent-child hierarchies, where each entity references its parent, defining the hierarchical structure. A common example is company organization structure or HR reporting, where each employee entity references another employee a as direct report or manager.
```

##### Domain Model

The simplest domain model looks as follows:

> ```cds
> entity Employee : Hierarchy {
>     key ID     : UUID;
>         parent : Association to Employee;
>         fullName   : String;
> }
> 
> aspect Hierarchy {
>   virtual LimitedDescendantCount : Integer64;
>   virtual DistanceFromRoot       : Integer64;
>   virtual DrillState             : String;
>   virtual LimitedRank            : Integer64;
> }
> ```

Reviewer: ```suggestion
```cds
entity Employee : Hierarchy {
  key ID       : UUID;
      parent   : Association to Employee;
      fullName : String;
}

aspect Hierarchy {
  virtual LimitedDescendantCount : Integer64;
  virtual DistanceFromRoot       : Integer64;
  virtual DrillState             : String;
  virtual LimitedRank            : Integer64;
}
```
```

> The entity `Employee` contains the element `ID`, which identifies the node, and the `parent` association referencing the same entity. Thus the parent-child relationship is set.

Reviewer: ```suggestion
The entity `Employee` has the element `ID`, which identifies the node. The `parent` association references the same entity, which establishes the parent-child relationship.
```

> The `Hierarchy` aspect defines a set of virtual elements, automatically calculated by the backend at runtime, to describe the state of the hierarchy. This information is requested by the UI to correctly render the hierarchy in a *TreeTable* during user interaction.

Reviewer: ```suggestion
##### Virtual Elements of a Hierarchy

The `Hierarchy` aspect defines a set of virtual elements, automatically calculated by the backend at runtime, to describe the state of the hierarchy. This information is requested by the UI to correctly render the hierarchy in a *TreeTable* during user interaction.
```

##### Service Model

The service defines the projection on the domain model.

```cds
@odata.apply.transformations
service HRService {
    entity HREmployee as projection on Employee;
}
```

::: warning

> The service must be annotated with `@odata.apply.transformations`. This instructs the Java Runtime to push down the whole transformation pipeline to the persistence service.

Reviewer: ```suggestion
The service must be annotated with `@odata.apply.transformations`. This instructs the Java Runtime to push down the whole transformation pipeline to the persistence layer.
```

   /ancestors(..., ID, filter(contains(Name, 'New York')), keep start)
```

#### Modeling Recursive Hierarchies

Recursive are parent-child hierarchies, where each entity references its parent, defining the hierarchical structure. A common example is company organization structure or HR reporting, where each employee entity references another employee a as direct report or manager.

##### Domain Model

The simplest domain model looks as follows:

```cds
entity Employee : Hierarchy {
  key ID       : UUID;
      parent   : Association to Employee;
      fullName : String;
}

aspect Hierarchy {
  virtual LimitedDescendantCount : Integer64;
  virtual DistanceFromRoot       : Integer64;
  virtual DrillState             : String;
  virtual LimitedRank            : Integer64;
}

> 

Author: ```suggestion
```

```

   /ancestors(..., ID, filter(contains(Name, 'New York')), keep start)
```

#### Modeling Recursive Hierarchies

> Recursive are parent-child hierarchies, where each entity references its parent, defining the hierarchical structure. A common example is company organization structure or HR reporting, where each employee entity references another employee a as direct report or manager.

Reviewer 2: ```suggestion
Recursive hierarchies are parent-child hierarchies, where each entity references its parent and through that defines the hierarchical structure. A common example is a company organization structure or HR reporting, where each employee entity references another employee a as direct report or manager.
```

##### Domain Model

The simplest domain model looks as follows:

```cds
entity Employee : Hierarchy {
  key ID       : UUID;
      parent   : Association to Employee;
      fullName : String;
}

aspect Hierarchy {
  virtual LimitedDescendantCount : Integer64;
  virtual DistanceFromRoot       : Integer64;
  virtual DrillState             : String;
  virtual LimitedRank            : Integer64;
}
```

> The entity `Employee` has the element `ID`, which identifies the node. The `parent` association references the same entity, which establishes the parent-child relationship.

Reviewer 2: Later one we call it hierarchy node. Use it also here to make it easier to find the connection?

Reviewer 2: ```suggestion
The entity `Employee` has the element `ID`, which identifies the hierarchy node. The `parent` association references the same entity, which establishes the parent-child relationship.
```

##### Virtual Elements of a Hierarchy

The `Hierarchy` aspect defines a set of virtual elements, automatically calculated by the backend at runtime, to describe the state of the hierarchy. This information is requested by the UI to correctly render the hierarchy in a *TreeTable* during user interaction.

##### Service Model

> The service defines the projection on the domain model.

Reviewer 2: ```suggestion
The following service defines the projection on the domain model.
```

```cds
@odata.apply.transformations
service HRService {
    entity HREmployee as projection on Employee;
}
```

::: warning
The service must be annotated with `@odata.apply.transformations`. This instructs the Java Runtime to push down the whole transformation pipeline to the persistence layer.
:::

##### OData v4 Annotations for Fiori

> In order to link the backend and Fiori UI, the projected service entity must be enriched with the following annotations.

Reviewer 2: ```suggestion
To link the backend and Fiori UI, the projected service entity must be enriched with the following annotations.
```
